Refactored RequestForm as a reusable component.

metincansiper opened the pull request:
The first phase of implementation for #901 as we talked with @maxkfranz.

maxkfranz commented:
Looks great.  The next step would be to add a 'Name' text field in the form that sets the contributing author's name.

metincansiper commented:
>Looks great. The next step would be to add a 'Name' text field in the form that sets the contributing author's name.

@maxkfranz I added that to this PR. I control the author name through ``doc.authorName()``. I am not sure if the function name is convenient. I could update it if desired.

jvwong commented on src/client/components/request-form.js:
```diff
@@ -0,0 +1,177 @@
+    const { paperId, authorEmail, authorName } = this.state;
+    const { checkIncomplete = false } = this.props;
+
+    if( !checkIncomplete && ( !paperId || !authorEmail || !authorName ) ){
```
This would imply that author name is required in order for the form to be complete. Is this correct? 
It was always optional as it was part of the post-submit case.

metincansiper replied:
@jvwong ``checkIncomplete`` is a parameter for this component and so whether to check if any field exists is controlled by that option. Currently either all the fields can be required or optional all together controlled by ``checkIncomplete`` parameter. You mean author name should still be different then other 2 and should not be required even when ``checkIncomplete``  parameter is set?

jvwong replied:
Oh OK that's fine.

jvwong commented on src/model/document/document.js:
```diff
@@ -455,6 +455,17 @@ class Document {
     }
   }
 
+  authorName( name ){
```
So my post-submit code stashes the 'name'  under `doc: { provided: { name: "xyz" }`. If you wanna keep this as a top-level attribute doc, you'll have to update the post-submit field:

https://github.com/PathwayCommons/factoid/blob/3a993ef5b7b6a9922a8f9d4053e29bcca0fdf124/src/client/components/tasks.js#L413

metincansiper replied:
@jvwong would this commit (https://github.com/PathwayCommons/factoid/pull/913/commits/4b2c3667d738e9d0ee8a83ca0f90c230cd517ec0) covers what you mean?

maxkfranz replied:
I agree that the data should be stored in the same way as before, ideally through the same PATCH route.

jvwong replied:
> @jvwong would this commit ([4b2c366](https://github.com/PathwayCommons/factoid/commit/4b2c3667d738e9d0ee8a83ca0f90c230cd517ec0)) covers what you mean?

My doc attribute is `provided.name` rather than `authorName`

metincansiper replied:
Okay sorry! I set ``provided.name`` instead of ``provided.authorName``.

metincansiper commented:
I realized that I implemented the logic of ``checkIncomplete`` parameter in opposite way and fixed it (https://github.com/PathwayCommons/factoid/pull/913/commits/157b3158f5c1e4807bc75b23428d2df4c7077dd8).

@maxkfranz I guess the next thing to implement is reusing the popover on the editor (around the title)? Actually I think the title will not be visible in that case. Therefore, I can just put the edit button in place of the title?

maxkfranz commented:
> @maxkfranz I guess the next thing to implement is reusing the popover on the editor (around the title)? Actually I think the title will not be available in that case. Therefore, I can just put the edit button in place of the title?

Yes, but in order to do that you need an easy way to create new documents without a title.  You could allow the admin interface to do this by modifying the form to allow the text boxes to be empty.  This could be an option in the component.

Once that's done, you can show some text like 'Set the article's title' where the title would normally go -- in the case where the document is empty.  Clicking that text would open the form, as you've described.

metincansiper commented:
>Yes, but in order to do that you need an easy way to create new documents without a title. You could allow the admin interface to do this by modifying the form to allow the text boxes to be empty. This could be an option in the component.

I think this is what the existing ``checkIncomplete`` option of component does. I am able to do it by using that option.

metincansiper commented:
@maxkfranz @jvwong I updated this PR by reusing the request for around the editor title. For now I enabled any field in the request form in the editor no matter if any of the fields is already set. I guess we may want to update the logic there. I realized that the server service for editing documents ('api/document/${id}/${secret}') checks ``document.provided()`` for paperId and authorEmail rather then the service parameter values. Therefore, I needed to update the ``document.provided()`` first and them make the service call as here (https://github.com/PathwayCommons/factoid/pull/913/commits/6d434488f756826389492a78d829b66c408258d2#diff-2e21dfae6311a798a1ba99795085d0176dba36a998c1f785963426bf83c2fbacR145). I wonder if it would be a inconsistency problem if there would be an exception between ``document.provided()``  and service calls? Also for the ``authorName`` I am just updating ``provided.name`` field. Should that one also be considered in the service like the other 2 fields?

maxkfranz commented:
Looks good.  I think it may be a bit simpler and less error prone for the case that you mentioned to just use the `doc` functions:

https://github.com/PathwayCommons/factoid/commit/6d434488f756826389492a78d829b66c408258d2#diff-2e21dfae6311a798a1ba99795085d0176dba36a998c1f785963426bf83c2fbacR113-R152

The extra fetch that you have at the end of that `else` is fine for now, but if the fetch/patch logic changes in future -- and it's changed many times already -- there could be a conflict.  If there are any missing setters in the `Document` class, feel free to add them to simplify the `else` block you mentioned.

So instead of making everything in the component use `fetch()`, that particular case can just use `doc` functions.  Does that make sense?

Once that's done, I think we can merge this in.

jvwong commented on src/client/components/editor/title.js:
```diff
@@ -1,28 +1,77 @@
+    }
+
+    const getTitleContent = () => {
+       if ( title == null ) {
```
What are the conditions under which I'm allowed to edit title? I don't think `null` is even possible?

maxkfranz replied:
It might be better as `if(title)` to handle the empty string case.  It should be possible to have an empty title as a part of the email campaign.  We send a completely empty document to a user so that they can try it out and fill in the paper details after the fact (#901).

jvwong replied:
I'm missed this discussion  but I thought the point was to always make the info editable?

maxkfranz replied:
The main editable case is for the when you start with an empty doc (e.g. related papers emails).  Is there another case that would benefit from editing that info after the fact?  If all the other cases have the info already set, is there a benefit of being able to change those values?

I suppose you could have an author's paper who changed the title of the paper slightly, so they want to update the factoid.  That's probably a rare edge case.

metincansiper replied:
I actually set the default value for the title as ``null`` instead of ``Untitled`` to better understand if it is actually set (https://github.com/PathwayCommons/factoid/pull/913/files#diff-bbf58575632e23c789ad93390b6495ff736ac319e4ee5e2473416a83e1bd03caR230).

jvwong commented:
Do we want the 'name' field in the homepage popup as a requirement?

maxkfranz commented:
> Do we want the 'name' field in the homepage popup as a requirement?

It helps to simplify things.  I've got some changes planned around the submission UI and this title editing popover planned after Metin's PR.

metincansiper commented:
>Looks good. I think it may be a bit simpler and less error prone for the case that you mentioned to just use the doc functions:
6d43448#diff-2e21dfae6311a798a1ba99795085d0176dba36a998c1f785963426bf83c2fbacR113-R152
The extra fetch that you have at the end of that else is fine for now, but if the fetch/patch logic changes in future -- and it's changed many times already -- there could be a conflict. If there are any missing setters in the Document class, feel free to add them to simplify the else block you mentioned.
So instead of making everything in the component use fetch(), that particular case can just use doc functions. Does that make sense?
Once that's done, I think we can merge this in.

@maxkfranz I see that the service that I am fetching is calling ``getPubmedArticle()`` which is also in server side. I do not know yet but there may be other similar things as well. Therefore, I think that I may not be able to put the same functionality under the ``Document`` model. What do you think?

maxkfranz commented:
That's a good point.  Let's go with the PR as-is, and we can reevaluate those details later if need be.
